### DISCUSSION TITLE
Skipping deploys programatically

### DISCUSSION THREAD

**Discussion by rodrigogs** [serverless/serverless#12145] [Q&A]
Is there a way to create a plugin to skip deployments as soon as possible?

I tried the following, but the `shouldNotDeploy` variable is reseted by the `aws:deploy:deploy:checkForChange` hook.

```javascript
export default class ServerlessSkip {
  constructor (serverless) {
    this.serverless = serverless

    this.hooks = {
      'package:initialize': this.skip.bind(this),
    }
  }

  skip () {
    this.serverless.cli.log('Skipping deployment')
    this.serverless.service.provider.shouldNotDeploy = true
  }
}
```

I want to create a plugin that checks if last git commit had changes to the service, otherwise I want to skip the deployment ASAP.

**Comment by SunitaMArshall** [comment]
Creating a plugin to skip deployments as soon as possible in the Serverless Framework can be a bit tricky because the framework itself is designed to follow a specific deployment process, which includes multiple hooks and stages. To achieve your goal of skipping the deployment as soon as possible based on whether there are changes in the last git commit, you can follow these steps:

Install the required npm packages for Git interaction and make sure you have git installed on your system.
npm install simple-git
Create a custom plugin to check for changes in the last Git commit and skip deployment if there are no changes.

const simpleGit = require('simple-git/promise');
const path = require('path');

class ServerlessSkipPlugin {
  constructor(serverless, options) {
    this.serverless = serverless;
    this.options = options;

    this.commands = {
      skip: {
        usage: 'Skip deployment if there are no changes in the last Git commit',
        lifecycleEvents: ['skipDeployment'],
      },
    };

    this.hooks = {
      'skip:skipDeployment': this.skipDeployment.bind(this),
      'before:package:initialize': this.beforePackageInitialize.bind(this),
    };
  }

  async skipDeployment() {
    try {
      const git = simpleGit(path.resolve(this.serverless.config.servicePath));
      const status = await git.status();
      
      if (status.files.length === 0) {
        this.serverless.cli.log('No changes in the last Git commit. Skipping deployment.');
        this.serverless.service.provider.shouldNotDeploy = true;
      } else {
        this.serverless.cli.log('Changes found in the last Git commit. Proceeding with deployment.');
      }
    } catch (err) {
      this.serverless.cli.log(`Error checking Git status: ${err.message}`);
      throw err;
    }
  }

  beforePackageInitialize() {
    if (this.serverless.service.provider.shouldNotDeploy) {
      this.serverless.service.provider.shouldNotDeploy = false;
    }
  }
}

module.exports = ServerlessSkipPlugin;

Modify your serverless.yml to include the plugin and add a custom command to skip the deployment.
service: my-service
provider:
  name: aws
  runtime: nodejs14.x

plugins:
  - serverless-skip-plugin

custom:
  shouldNotDeploy: false # Initialize the flag to false

functions:
  # ...

# Add this custom command to your serverless.yml
custom:
  shouldNotDeploy: false
  serverless-skip-plugin:
    # Configuration for the plugin (if needed)

Now, you can use the serverless skip command to check for changes in the last Git commit and skip the deployment if no changes are found.
serverless skip
This custom plugin should help you achieve your goal of skipping the deployment as soon as possible based on the Git commit status.

**Reply by rodrigogs** [reply]
Wow, I really appreciate your solution. However, I do not wish to create a new command for serverless, because it would end up breaking the serverless compose (which I'm using).

Should I perhaps open a feature request to be able to programmatically skip a deploy? I really think it would be a great feature.
